README.md
# NativeScript Calendar Plugin

[![Build Status][build-status]][build-url]
[![NPM version][npm-image]][npm-url]
[![Downloads][downloads-image]][npm-url]
[![Twitter Follow][twitter-image]][twitter-url]

[build-status]:https://travis-ci.org/EddyVerbruggen/nativescript-calendar.svg?branch=master
[build-url]:https://travis-ci.org/EddyVerbruggen/nativescript-calendar
[npm-image]:http://img.shields.io/npm/v/nativescript-calendar.svg
[npm-url]:https://npmjs.org/package/nativescript-calendar
[downloads-image]:http://img.shields.io/npm/dm/nativescript-calendar.svg
[twitter-image]:https://img.shields.io/twitter/follow/eddyverbruggen.svg?style=social&label=Follow%20me
[twitter-url]:https://twitter.com/eddyverbruggen

The Calendar plugin allows you to manipulate events in the user's native Calendar. You can find, create and delete events in either the default or a custom calendar.

If you're looking for an awesome in-app UI for the native calendar, then [check this out](http://docs.telerik.com/devtools/nativescript-ui/Controls/NativeScript/Calendar/overview).

## Installation
From the command prompt go to your app's root folder and execute:
```
tns plugin add nativescript-calendar
```

## iOS runtime permission reason
You probably have seen a permission popup like this before (this plugin will trigger one as well, automatically):

<img src="doc/ios_permission_custom_reason.png" width="316px" height="180px"/>

iOS 10+ requires not only this popup, but also a _reason_. In this case it's  "Custom message from App_Resources".

You can provide your own reason for accessing the calendar by adding something like this to `app/App_Resources/ios/Info.plist`:

```xml
  <key>NSCalendarsUsageDescription</key>
  <string>My reason justifying fooling around with your calendar</string>
```

_To not crash your app in case you forgot to provide the reason this plugin adds an empty reason to the `.plist` during build. This value gets overridden by anything you specified yourself. You're welcome._

## TypeScript Usage
Of course you can use this plugin with TypeScript, just import the plugin and use
the functions summed up below like this:

```typescript
import * as Calendar from "nativescript-calendar";

// example for listCalendars:
Calendar.listCalendars().then(/* .. */);
```

## Usage

If you want a quickstart, [clone our demo app](https://github.com/EddyVerbruggen/nativescript-calendar-demo).

### createEvent

```js
  var Calendar = require("nativescript-calendar");

  // Only the `title`, `startDate` and `endDate` are mandatory, so this would suffice:
  var options = {
    title: 'Get groceries',
    // Make sure these are valid JavaScript Date objects.
    // In this case we schedule an Event for now + 1 hour, lasting 1 hour.
    startDate: new Date(new Date().getTime() + (60*60*1000)),
    endDate: new Date(new Date().getTime() + (2*60*60*1000))
  };

  // You can however add lots of properties to enrich the Event:
  options.location = 'The shop';
  options.notes = 'This event has reminders';

  // iOS has a separate 'url' field, but on Android the plugin appends this to the 'notes' field.
  options.url = 'http://my.shoppinglist.com';

  // You can also override the default reminder(s) of the Calendar (in minutes):
  options.reminders = {
    first: 30,
    second: 10
  };

  // You can make this Event recurring (this one repeats every other day for 10 days):
  options.recurrence = {
    frequency: "daily", // daily | weekly | monthly | yearly
    interval: 2, // once every 2 days
    endDate: new Date(new Date().getTime() + (10*24*60*60*1000)) // 10 days
  };

  // Want to use a custom calendar for your app? Pass in the 'name'.
  // If the name doesn't yet exist the plugin will create it for you.
  options.calendar = {
    name: "NativeScript Cal",
    // the color, in this case red
    color: "#FF0000",
    // Can be used on Android to group the calendars. Examples: Your app name, or an emailaddress
    accountName: "My App Name"
  };

  Calendar.createEvent(options).then(
      function(createdId) {
        console.log("Created Event with ID: " + createdId);
      },
      function(error) {
        console.log("Error creating an Event: " + error);
      }
  );
```

If you want an 'all day event', make sure you set the dates to midnight like this:
```js
  var d = new Date();
  d.setHours(0);
  d.setMinutes(0);
  d.setSeconds(0);

  // this will create an 'all day event' for tomorrow
  var startDate = new Date(d.getTime() + (24*60*60*1000));
  var endDate = new Date(d.getTime() + (2*24*60*60*1000));
  // .. now use these properties in the options object
```

### findEvents
```js
  var options = {
    // when searching, dates are mandatory - the event must be within this interval
    startDate: new Date(new Date().getTime() - (50*24*60*60*1000)),
    endDate: new Date(new Date().getTime() + (50*24*60*60*1000))
  };

  // if you know the Event ID, set it here:
  options.id = '123456';

  // you can optionally pass in a few other properties, any event containing these will be returned:
  options.title = 'groceries';
  options.location = 'foo';
  options.notes = 'bar'; // iOS only

  Calendar.findEvents(options).then(
      function(events) {
        console.log(JSON.stringify(events));
      },
      function(error) {
        console.log("Error finding Events: " + error);
      }
  );
```

The returned 'events' object is an array of JSON events with these properties:
```js
id
title
location
notes
url
startDate
endDate
allDay
calendar {id, name}
reminders {minutes}
recurrence {frequency, interval, endDate}
attendees {name, url, status, role, type} // iOS only
```

### deleteEvents
Usage is largely the same as findEvents, only the result is a bit different ;)

```js
  var options = {
    // when searching, dates are mandatory - the event must be within this interval
    startDate: new Date(new Date().getTime() - (50*24*60*60*1000)),
    endDate: new Date(new Date().getTime() + (50*24*60*60*1000))
  };

  // if you know the Event ID, set it here:
  options.id = '123456';

  // you can optionally pass in a few other properties, any event containing these will be deleted:
  options.title = 'groceries'; // events _including_ this string will be included in the selection
  options.location = 'foo';
  options.notes = 'bar'; // iOS only

  Calendar.deleteEvents(options).then(
      function(deletedEventIds) {
        console.log(JSON.stringify(deletedEventIds));
      },
      function(error) {
        console.log("Error deleting Events: " + error);
      }
  )
```

### listCalendars
```js
  Calendar.listCalendars().then(
      function(calendars) {
        // a JSON array of Calendar objects is returned, each with an 'id' and 'name'
        console.log("Found these Calendars on the device: " + JSON.stringify(calendars));
      },
      function(error) {
        console.log("Error while listing Calendars: " + error);
      }
  )
```

### deleteCalendar

##### TypeScript
```typescript
import * as Calendar from "nativescript-calendar";

Calendar.deleteCalendar({
  name: "My Calendar name"
}).then(id => {
  // id is null if nothing was deleted
  console.log(`Deleted Calendar with id ${id}`);
});
```

## Breaking changes in 2.0.0
See [CHANGELOG.md](CHANGELOG.md).


CHANGELOG.md
# Change Log

## [2.0.0](https://github.com/EddyVerbruggen/nativescript-calendar/tree/2.0.0) (2018-02-26)
[Full Changelog](https://github.com/EddyVerbruggen/nativescript-calendar/compare/1.3.0...2.0.0)

**BREAKING CHANGES:**
- The `recurrence.frequency` property of `createEvent` changed to a string instead of a constant. Use one of `daily | weekly | monthly | yearly`.
- The `recurringRule` property of `findEvents` has been removed (it was never set for iOS anyway). Replaced by the new `recurrence` object, which is available on both iOS and Android.

**Implemented enhancements:**

- How to get recurrence value on findEvents method? [\#26](https://github.com/EddyVerbruggen/nativescript-calendar/issues/26)
- Event/Calendar ID's must be of type 'string' [\#39](https://github.com/EddyVerbruggen/nativescript-calendar/issues/39)
- Add 'reminders' to 'findEvents' [\#40](https://github.com/EddyVerbruggen/nativescript-calendar/issues/40)

**Fixed bugs:**

- I could not get notes, rrule, begin, end values on ios! [\#31](https://github.com/EddyVerbruggen/nativescript-calendar/issues/31)
- (Events in) custom calendars on Android can't be viewed nor removed [\#42](https://github.com/EddyVerbruggen/nativescript-calendar/issues/42)
- Deleting events on iOS only works by 'id' [\#43](https://github.com/EddyVerbruggen/nativescript-calendar/issues/43)

## [1.3.0](https://github.com/EddyVerbruggen/nativescript-calendar/tree/1.3.0) (2018-01-09)
[Full Changelog](https://github.com/EddyVerbruggen/nativescript-calendar/compare/1.2.3...1.3.0)

**Implemented enhancements:**

- Align with NativeScript seed (inline demo app, convert to TypeScript, etc) [\#34](https://github.com/EddyVerbruggen/nativescript-calendar/issues/34)
- Add displayname to android calendars [\#29](https://github.com/EddyVerbruggen/nativescript-calendar/pull/29)


## [1.2.3](https://github.com/EddyVerbruggen/nativescript-calendar/tree/1.2.3) (2018-01-09)
[Full Changelog](https://github.com/EddyVerbruggen/nativescript-calendar/compare/1.2.2...1.2.3)

**Fixed bugs:**

- App crash when denying permissions [\#33](https://github.com/EddyVerbruggen/nativescript-calendar/issues/33)


## [1.2.2](https://github.com/EddyVerbruggen/nativescript-calendar/tree/1.2.2) (2017-05-05)
[Full Changelog](https://github.com/EddyVerbruggen/nativescript-calendar/compare/1.2.1...1.2.2)

**Fixed bugs:**

- Problem with building an app with the plugin for Android [\#22](https://github.com/EddyVerbruggen/nativescript-calendar/issues/22)

## [1.2.1](https://github.com/EddyVerbruggen/nativescript-calendar/tree/1.2.1) (2017-02-11)
[Full Changelog](https://github.com/EddyVerbruggen/nativescript-calendar/compare/1.2.0...1.2.1)

**Fixed bugs:**

- Possible to specify no reminder? [\#15](https://github.com/EddyVerbruggen/nativescript-calendar/issues/15)

## [1.2.0](https://github.com/EddyVerbruggen/nativescript-calendar/tree/1.2.0) (2016-11-22)
[Full Changelog](https://github.com/EddyVerbruggen/nativescript-calendar/compare/1.1.3...1.2.0)

**Implemented enhancements:**

- New feature: allow specifying a calendar color [\#14](https://github.com/EddyVerbruggen/nativescript-calendar/issues/14)
- New feature: deleteCalendar [\#13](https://github.com/EddyVerbruggen/nativescript-calendar/issues/13)
- Android: Cannot delete custom local calendar [\#6](https://github.com/EddyVerbruggen/nativescript-calendar/issues/6)

**Fixed bugs:**

- Android: Calendar.listCalendars \(actually Calendar.\_findCalendars\) results in a crash [\#12](https://github.com/EddyVerbruggen/nativescript-calendar/issues/12)
- Android: Cannot delete custom local calendar [\#6](https://github.com/EddyVerbruggen/nativescript-calendar/issues/6)

**Closed issues:**

- findEvents logs an error but returns the correct result \(iOS Simulator?\) [\#11](https://github.com/EddyVerbruggen/nativescript-calendar/issues/11)
- How to use this in ng nativescript [\#8](https://github.com/EddyVerbruggen/nativescript-calendar/issues/8)

## [1.1.3](https://github.com/EddyVerbruggen/nativescript-calendar/tree/1.1.3) (2016-10-29)
[Full Changelog](https://github.com/EddyVerbruggen/nativescript-calendar/compare/1.1.2...1.1.3)

**Fixed bugs:**

- Assertion failure [\#10](https://github.com/EddyVerbruggen/nativescript-calendar/issues/10)

## [1.1.2](https://github.com/EddyVerbruggen/nativescript-calendar/tree/1.1.2) (2016-10-29)
[Full Changelog](https://github.com/EddyVerbruggen/nativescript-calendar/compare/1.1.1...1.1.2)

**Fixed bugs:**

- \[Android\] Calendar.findEvents missing id-filter [\#9](https://github.com/EddyVerbruggen/nativescript-calendar/issues/9)

## [1.1.1](https://github.com/EddyVerbruggen/nativescript-calendar/tree/1.1.1) (2016-10-15)
[Full Changelog](https://github.com/EddyVerbruggen/nativescript-calendar/compare/1.0.7...1.1.1)

**Implemented enhancements:**

- Feature Suggestion: Check to see if the user has granted permission to access the calendar [\#1](https://github.com/EddyVerbruggen/nativescript-calendar/issues/1)

**Fixed bugs:**

- Missing comma in 1.1.0 preventing it from installing [\#7](https://github.com/EddyVerbruggen/nativescript-calendar/issues/7)
- NPE on Android in \_requestPermission executed by timeout [\#5](https://github.com/EddyVerbruggen/nativescript-calendar/issues/5)

**Closed issues:**

- Android - how to get the permissions prompt [\#3](https://github.com/EddyVerbruggen/nativescript-calendar/issues/3)

## [1.0.7](https://github.com/EddyVerbruggen/nativescript-calendar/tree/1.0.7) (2016-08-22)
[Full Changelog](https://github.com/EddyVerbruggen/nativescript-calendar/compare/1.0.6...1.0.7)

**Fixed bugs:**

- \[Android\] Not setting options.url adds undefined to description [\#2](https://github.com/EddyVerbruggen/nativescript-calendar/issues/2)

## [1.0.6](https://github.com/EddyVerbruggen/nativescript-calendar/tree/1.0.6) (2016-06-27)
[Full Changelog](https://github.com/EddyVerbruggen/nativescript-calendar/compare/1.0.5...1.0.6)


src/index.d.ts
declare module "nativescript-calendar" {

  export type RecurrenceFrequency = "daily" | "weekly" | "monthly" | "yearly";
  export type CalendarType = "Local" | "CalDAV" | "Exchange" | "Subscription" | "Birthday" | "Mail";

  interface Recurrence {
    frequency: RecurrenceFrequency;
    /**
     * Default 1 (every <RecurrenceFrequency>).
     */
    interval: number;
    endDate?: Date;
    count?: number;
  }

  /**
     * The options object passed into the createEvent function.
     */
    export interface CreateEventOptions {
      /**
       * The title of the event.
       */
      title: string;

      /**
       * A valid JS Date representing the start date (and time of the event).
       * If you want an 'All day event', make sure you set the dates to midnight like this:
       *
       *    var d = new Date();
       *    d.setHours(0);
       *    d.setMinutes(0);
       *    d.setSeconds(0);
       *
       *    // then this will create an 'all day event' for tomorrow
       *    startDate = new Date(d.getTime() + (24*60*60*1000));
       *    endDate = new Date(d.getTime() + (2*24*60*60*1000));
       */
      startDate: Date;

      /**
       * A valid JS Date representing the end date (and time of the event).
       */
      endDate: Date;

      /**
       * Where the event takes place.
       */
      location?: string;

      /**
       * Any remarks you want to store with the event.
       */
      notes?: string;

      /**
       * On iOS there's a seperate field for storing a URL with the event.
       * On Android there's not, so we add it to any notes you pass in automatically.
       */
      url?: URL;

      /**
       * Want to use a custom calendar for your app? Pass in the 'name'.
       * If the name doesn't yet exist the plugin will create it for you.
       */
      calendar: {
        name: string;
        /**
         * Example, red: "#FF0000"
         */
        color?: string;
        /**
         * Can be used on Android to group the calendars.
         * Examples: Your app name, or an emailaddress.
         */
        accountName?: string;
      };

      /**
       * Override the default reminders if you like.
       * If you don't the plugin will set a reminder of 60 minutes before the event automatically.
       */
      reminders?: {
        /**
         * Set to null if you don't want a reminder at all.
         * Default 60 (minutes).
         */
        first?: number;
        /**
         * Default null (no second reminder).
         */
        second?: number;
      };

      /**
       * Use this if you want this event to repeat with a certain interval.
       * For instance, if you want an event to recur every other day for 10 days, use:
       * {
       *   frequency: "daily" | "weekly" | "monthly" | "yearly",
       *   interval: 2,
       *   endDate: new Date(new Date().getTime() + (10*24*60*60*1000))
       * }
       */
      recurrence?: Recurrence;
    }

    interface FindOrDeleteEventsOptions {
      /**
       * When searching, dates are mandatory - the event must be within this interval.
       */
      startDate: Date;

      /**
       * When searching, dates are mandatory - the event must be within this interval.
       */
      endDate: Date;

      /**
       * If you know the Event ID, set it here.
       */
      id?: string;

      /**
       * (Part of) the title of the event.
       */
      title?: string;

      /**
       * (Part of) the location of the event.
       */
      location?: string;

      /**
       * (Part of) the notes of the event.
       * iOS only.
       */
      notes?: string;
    }

    export interface FindEventsOptions extends FindOrDeleteEventsOptions {
    }

    export interface DeleteEventsOptions extends FindOrDeleteEventsOptions {
    }

    export interface DeleteCalendarOptions {
      name: string;
    }

    export interface Calendar {
      id: string;
      name: string;
      /**
       * iOS: same as 'name'
       * Android: usually the same as 'name' as well
       */
      displayName?: string;
      /**
       * iOS only
       */
      type?: CalendarType;
    }

    export interface Event {
      id: string;
      title: string;
      location: string;
      notes: string;
      url?: string;
      startDate: Date;
      endDate: Date;
      allDay: boolean;
      calendar: Calendar;
      recurrence?: Recurrence;
      reminders?: Array<Reminder>;
      /**
       * iOS only.
       */
      attendees?: Array<Attendee>;
    }

    export interface Attendee {
      name: string;
      url?: string;
      /**
       * One of: "Unknown", "Pending", "Accepted", "Declined", "Tentative", "Delegated", "Completed", "In Process"
       */
      status: string;
      /**
       * One of: "Unknown", "Required", "Optional", "Chair", "Non Participant"
       */
      role: string;
      /**
       * One of: "Unknown", "Person", "Room", "Resource", "Group"
       */
      type: string;
    }

    export interface Reminder {
      minutes: number;
    }

    /**
     * Returns the ID of the event that was created.
     */
    export function createEvent(options: CreateEventOptions): Promise<string>;

    /**
     * Find events matched on ALL params passed in.
     */
    export function findEvents(options: FindEventsOptions): Promise<Event[]>;

    /**
     * Usage is the same as findEvents, but the result is a bit different ;)
     * Returns an array of deleted event ID's.
     */
    export function deleteEvents(options: DeleteEventsOptions): Promise<string[]>;

    /**
     * List all available Calendars on the user's device.
     */
    export function listCalendars(): Promise<Calendar[]>;

    /**
     * Returns the ID of the deleted calendar (or null if none was deleted).
     */
    export function deleteCalendar(options: DeleteCalendarOptions): Promise<string>;

    /**
     * No real reason to use this as it's all handled automatically for you.
     */
    export function hasPermission(): Promise<boolean>;

    /**
     * No real reason to use this as it's all handled automatically for you.
     */
    export function requestPermission(): Promise<any>;
}
